### ISSUE TITLE
How to Set the Locale in metadata and generate postcodes 

### ISSUE THREAD

**Issue by JahangirM** [sdv-dev/SDV#1370]
### Environment details

If you are already running SDV, please indicate the following details about the environment in
which you are running it:

* SDV version: v1.0.0
* Python version: 3.10.10
* Operating System: Windows 11

### Problem description

Hello fairly new to SDV and I am trying to generate postcodes for the UK. I am trying to use the `en_GB` locale provider https://faker.readthedocs.io/en/master/locales/en_GB.html.

I am running into the following issues:
- Struggling to set the locale provider via the metadata json
- When generating postcodes (even using en_US locale) I get the below error
```
Traceback (most recent call last):
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\rdt\transformers\pii\anonymizer.py", line 54, in check_provider_function
    module = getattr(faker.providers, provider_name)
AttributeError: module 'faker.providers' has no attribute 'en_GB'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Repos\SyntheticDataGAN\DataGenerator\main.py", line 45, in <module>
    main()
  File "C:\Repos\SyntheticDataGAN\DataGenerator\main.py", line 29, in main
    model = synthesizer.fit(df)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\single_table\base.py", line 456, in fit
    processed_data = self._preprocess(data)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\single_table\base.py", line 403, in _preprocess
    self._data_processor.fit(data)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\data_processing\data_processor.py", line 616, in fit
    self.prepare_for_fitting(data)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\data_processing\data_processor.py", line 579, in prepare_for_fitting
    config = self._create_config(constrained, columns_created_by_constraints)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\data_processing\data_processor.py", line 467, in _create_config
    transformers[column] = self.create_anonymized_transformer(
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\data_processing\data_processor.py", line 370, in create_anonymized_transformer
    return get_anonymized_transformer(sdtype, kwargs)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\metadata\anonymization.py", line 99, in get_anonymized_transformer
    return AnonymizedFaker(**function_kwargs)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\rdt\transformers\pii\anonymizer.py", line 100, in __init__
    self.check_provider_function(self.provider_name, self.function_name)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\rdt\transformers\pii\anonymizer.py", line 63, in check_provider_function
    raise TransformerProcessingError(
rdt.errors.TransformerProcessingError: The 'en_GB' module does not contain a function named 'postcode'.
Refer to the Faker docs to find the correct function: https://faker.readthedocs.io/en/master/providers.html
```

### What I already tried
Below is the metadata json I am using:
```
{
    "columns": {
        "name": {
            "sdtype": "name"
        },
        "age": {
            "sdtype": "numerical"
        },
        "sex": {
            "sdtype": "categorical"
        },
        "postcode": {
                    "sdtype": "postcode"

        }
    },
    "METADATA_SPEC_VERSION": "SINGLE_TABLE_V1"
}

```

I am currently able to set the locale with the following lines of code (taken from https://github.com/sdv-dev/SDV/issues/308):

```
import faker
#change locale
faker.proxy.DEFAULT_LOCALE = 'en_GB'
```

However I have come across the following docs which mention `pii_locales`:
- https://sdv.dev/SDV/developer_guides/sdv/metadata.html#localized-data-anonymization
- https://github.com/sdv-dev/SDV/discussions/667

I tried updating the metadata json to include these:
```
        "postcode": {
                    "sdtype": "postcode",
                    "pii": true,
                    "pii_locales": "en_GB"

        }
```

However I get the below error:
```
Traceback (most recent call last):
  File "C:\Repos\SyntheticDataGAN\DataGenerator\main.py", line 45, in <module>
    main()
  File "C:\Repos\SyntheticDataGAN\DataGenerator\main.py", line 26, in main
    synthesizer = GaussianCopulaSynthesizer(metadata)
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\single_table\copulas.py", line 93, in __init__
    super().__init__(
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\single_table\base.py", line 75, in __init__
    self.metadata.validate()
  File "C:\Repos\SyntheticDataGAN\venv\lib\site-packages\sdv\metadata\single_table.py", line 467, in validate
    raise InvalidMetadataError(
sdv.metadata.errors.InvalidMetadataError: The following errors were found in the metadata:

Invalid values '(pii_locales)' for postcode column 'postcode'.
```

I would like to know how I can set the locale within the metadata?

### Postcode Error

In addition to this when trying to generate postcode using the default locale `en_US`, I get the error mentioned in the **Problem Description** section. 

Looking at the faker providers I can see postcode for `en_US` and `en_GB` 
https://faker.readthedocs.io/en/master/locales/en_GB.html#faker.providers.address.en_GB.Provider.postcode

Below is an example of the code I am running:
```
from sdv.metadata import SingleTableMetadata
from sdv.single_table import GaussianCopulaSynthesizer

import pandas as pd

import faker

# change locale
faker.proxy.DEFAULT_LOCALE = "en_GB"


DATA = pd.DataFrame(
    data={
        "name": ["simon ross", "eliot lee"],
        "age": [22, 23],
        "sex": ["M", "F"],
        "postcode": ["xc12 3bq", "gd1 9ja"],
    }
)

METADATA = {
    "columns": {
        "name": {"sdtype": "name"},
        "age": {"sdtype": "numerical"},
        "sex": {"sdtype": "categorical"},
        "postcode": {"sdtype": "postcode"},
    },
    "METADATA_SPEC_VERSION": "SINGLE_TABLE_V1",
}


def main():
    metadata = SingleTableMetadata.load_from_dict(METADATA)
    # Step 1: Create the synthesizer
    synthesizer = GaussianCopulaSynthesizer(metadata)
    # Step 2: Train the synthesizer
    synthesizer.fit(DATA)  # <---- Error happens here

```

**Comment by npatki** [comment]
Hi @JahangirM, nice to meet you!

You can set the locales for each individual transformer by using the `auto_assign_transformers` and `update_transformers` functionality. We are actively working on improving this API. 

References:
- [This issue](https://github.com/sdv-dev/SDV/issues/1371) shows how you can set the locales today (and how we want to improve this!)
- [This demo notebook](https://colab.research.google.com/drive/1b0W9jpgOHFrkz-i1QynVuaMZXs7BSPb_?usp=sharing#scrollTo=A7I4ZYbT6tI5) shows you how anonymization generally works in the SDV

FYI We have recently introduced the new, SDV 1.0 library, the most up-to-date docs are found [here](https://docs.sdv.dev/sdv).  If you are referencing the [older docs](https://sdv.dev/SDV/developer_guides/sdv/metadata.html#localized-data-anonymization), the code will not work. One way to check is to scroll to the top of the page. If you see the following box, you are referencing out-of-date docs:

![image](https://user-images.githubusercontent.com/3227804/231551152-3d612413-7320-4381-a4de-f0548271c18f.png)

**Comment by JahangirM** [comment]
Thanks for the above. Able to confirm that I am able to set the locales per column which is great. 

One thing to note is that updating a transformer requires fitting first. When fitting using the sdtype `postcode` I run into the error mentioned above:
```
rdt.errors.TransformerProcessingError: The 'en_GB' module does not contain a function named 'postcode'.
Refer to the Faker docs to find the correct function: https://faker.readthedocs.io/en/master/providers.html
```

I have found a way around this by first changing the sdtypes from `postcode` to `address`, fitting the data and then updating the transformer. Example below:


## Original code that causes error
```
from rdt.transformers.pii import AnonymizedFaker
from sdv.metadata import SingleTableMetadata
from sdv.single_table import GaussianCopulaSynthesizer

import pandas as pd

DATA = pd.DataFrame(
    data={
        "name": ["simon ross", "eliot lee"],
        "age": [22, 23],
        "sex": ["M", "F"],
        "postcode": ["xc12 3bq", "gd1 9ja"],
    }
)

METADATA = {
    "columns": {
        "name": {"sdtype": "name"},
        "age": {"sdtype": "numerical"},
        "sex": {"sdtype": "categorical"},
        "postcode": {"sdtype": "postcode"},
    },
    "METADATA_SPEC_VERSION": "SINGLE_TABLE_V1",
}


def main():
    metadata = SingleTableMetadata.load_from_dict(METADATA)
    # Step 1: Create the synthesizer
    synthesizer = GaussianCopulaSynthesizer(metadata)

    # Step 2: initial fitting
    synthesizer.auto_assign_transformers(DATA) # <---- Error happens here

    synthesizer.update_transformers(
        column_name_to_transformer={
            "postcode": AnonymizedFaker(
                provider_name="address", function_name="postcode", locales=["en_GB"]
            )
        }
    )
    # Step 3: fit again
    synthesizer.fit(DATA)
    # Step 4: Generate synthetic data
    synthetic_data = synthesizer.sample(num_rows=100)
    print(synthetic_data)


if __name__ == "__main__":
    main()
```


## Working code where metadata sdtype is changed from postcode to address

```
from rdt.transformers.pii import AnonymizedFaker
from sdv.metadata import SingleTableMetadata
from sdv.single_table import GaussianCopulaSynthesizer

import pandas as pd


DATA = pd.DataFrame(
    data={
        "name": ["simon ross", "eliot lee"],
        "age": [22, 23],
        "sex": ["M", "F"],
        "postcode": ["xc12 3bq", "gd1 9ja"],
    }
)

METADATA = {
    "columns": {
        "name": {"sdtype": "name"},
        "age": {"sdtype": "numerical"},
        "sex": {"sdtype": "categorical"},
        "postcode": {"sdtype": "address"},  # <--- Change to address
    },
    "METADATA_SPEC_VERSION": "SINGLE_TABLE_V1",
}


def main():
    metadata = SingleTableMetadata.load_from_dict(METADATA)
    # Step 1: Create the synthesizer
    synthesizer = GaussianCopulaSynthesizer(metadata)

    # Step 2: initial fitting
    synthesizer.auto_assign_transformers(DATA)

    # Step 2a: Update the transformer so that it generate a postcode instead of address
    synthesizer.update_transformers(
        column_name_to_transformer={
            "postcode": AnonymizedFaker(
                provider_name="address", function_name="postcode", locales=["en_GB"]
            )
        }
    )
    # Step 3: fit again
    synthesizer.fit(DATA)
    # Step 4: Generate synthetic data
    synthetic_data = synthesizer.sample(num_rows=100)
    print(synthetic_data)


if __name__ == "__main__":
    main()

```

I am not sure if I am correctly generating postcodes or if this is a bug?

**Comment by npatki** [comment]
Hi @JahangirM, thanks for the details. What you are observing in the first section seems to be a bug in how we are setting the default localization. I've created #1372 to track this.

I see nothing wrong in your workaround! Even though you'd had to change the sdtype to `"address"`, you've correctly set the transformer after that so the synthetic data will be generating the correct postcodes.

**Comment by JahangirM** [comment]
Thanks based on #1372 looks like this should fixed in the next release. Thanks for the help!

**Comment by npatki** [comment]
Hi @JahangirM, this feature is now available in the [latest SDV release, 1.1.0](https://github.com/sdv-dev/SDV/releases/tag/v1.1.0).

You can now use the `locales` keyword to set the default locales upfront when creating the synthesizer. For eg.

```python
from sdv.single_table import GaussianCopulaSynthesizer

synthesizer = GaussianCopulaSynthesizer(
    metadata,
    locales=['en_GB']
)
```

For more info, see [the SDV docs](https://docs.sdv.dev/sdv/).
